base/linalg/triangular.jl

```diff
@@ -1369,34 +1369,217 @@ end
 At_ldiv_B(A::AbstractTriangular, B::AbstractVecOrMat) = At_ldiv_B!(A, copy(B))
 Ac_ldiv_B(A::AbstractTriangular, B::AbstractVecOrMat) = Ac_ldiv_B!(A, copy(B))
 
+function ^(A::UpperTriangular, p::Integer)
+    if p < 0
+        return UpperTriangular(Base.power_by_squaring(inv(A), -p))
+    else
+        return UpperTriangular(Base.power_by_squaring(A, p))
+    end
+end
+function ^(A::UpperTriangular, p::Real)
+    # For integer powers, use repeated squaring
+    if isinteger(p)
+        return A^Integer(real(p))
+    end
+
+    # General UpperTriangular matrices
+    p1 = p - floor(p)
+    if cond(A) >= exp(log(p1/(1 - p1)) / p1)
+        return powm(A,p1) * A^floor(p)
```

Comment:
Note that `floor(p)` has the same type as `p`. Here, `floor(Int, p)` should be better as it would trigger the integer version of `M^p` directly.